### ISSUE TITLE
Strange behaviour of <SuspenseList revealOrder="forwards"/> in concurrent mode

### ISSUE THREAD

**Issue by dubzzz** [facebook/react#17515]
_Behaviour looks pretty strange in my opinion but it might be the expected one. I just wanted to open an issue in case it was not. Please note that it is relative to the coming concurrent mode. Please feel free to close if it is not considered as a bug._

**What is the current behavior?**

_In concurrent mode_

The output of the following scenario looks strange to me. I'd have expected to have either `Loading A / Loading C / Loading C` or `Loading A / C / C`. (`A`, `B`, `C` are described right after)

Here is the scenario ([live demo with concurrent mode enabled](https://dubzzz.github.io/react-suspenselist-bug/build/)):

1. Render:
```jsx
render(
  <SuspenseList revealOrder="forwards">
    <Suspense fallback={<span>Loading B</span>}>
      <B />
    </Suspense>
    <Suspense fallback={<span>Loading C</span>}>
      <C />
    </Suspense>
  </SupenseList>
)
```

2. Resolve `B` and `C`

3. (Re-)render:
```jsx
render(
  <SuspenseList revealOrder="forwards">
    <Suspense fallback={<span>Loading A</span>}>
      <A />
    </Suspense>
    <Suspense fallback={<span>Loading C</span>}>
      <C />
    </Suspense>
    <Suspense fallback={<span>Loading C</span>}>
      <C />
    </Suspense>
  </SupenseList>
)
```

4. Output is: `Loading A / C / Loading C`

With:

```jsx
// You can considerer the following:
const A = React.lazy(() => import('./A'));
const B = React.lazy(() => import('./B'));
const C = React.lazy(() => import('./C'));
```
In my code: https://github.com/dubzzz/react-suspenselist-bug/blob/master/src/App.js#L189_L205

**What is the expected behavior?**

I'd have expected to have either `Loading A / Loading C / Loading C` or `Loading A / C / C`.

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**

React: `master` branch
React: `0.0.0-experimental-b53ea6ca0`

**How did I found this bug?**

This _potential bug_ has been discovered while I was trying to run property based tests based on [fast-check](https://github.com/dubzzz/fast-check/) against React library.

See https://github.com/dubzzz/react/commit/79d49ed85c34a5339f1d3163a0c6e99eddaf4405

In my test I was assessing the following properties:
- `"together"`: only renders when all components are ready (test covers complex and deep structures containing nested `<SuspenseList revealOrder="together" />` and `<Suspense />` referencing potentially multiple times the same `Component`)
- `"forwards"`: only renders components up to the last resolved one, all components after this one will be considered as unresolved

**Comment by gaearon** [comment]
>I'd have expected to have either Loading A / Loading C / Loading C or Loading A / C / C.

Can you elaborate on why?

One thing to note is that the list is not supposed to hide content that was already shown. 

It is also worth noting that keys matter. If you render two different JSX trees without keys then from React’s point of view, first item changed from B to A. If you wanted to express “A was inserted” you’d need to give them keys.

**Comment by dubzzz** [comment]
Thanks for the quick reply. The behaviour observed in the scenario described above is indeed expected. 👍 

I took into account your clarifications around what to expect from `React.SuspenseList` on re-renders to update my property based tests and re-ran them to see if I could find something else.

@gaearon Here is another strange scenario where I'm relying on keys so that React should not get lost (see https://dubzzz.github.io/react-suspenselist-bug/build/ for a live demo - [code here](https://github.com/dubzzz/react-suspenselist-bug/blob/master/src/App.js)):

1. Render:

```jsx
render(
  <SuspenseList key="1" revealOrder="forwards">
    <SuspenseList key="1.1" revealOrder="forwards">
      <Suspense key="1.1.c" fallback={<Text text="Loading C" />}>
        <C />
      </Suspense>
    </SuspenseList>
  </SuspenseList>
)
```

2. Resolve `C`

3. (Re-)render:

```jsx
  <SuspenseList key="1" revealOrder="forwards">
    <SuspenseList key="1.1" revealOrder="forwards">
      <Suspense key="1.1.a" fallback={<Text text="Loading A" />}>
        <A />
      </Suspense>
      <Suspense key="1.1.b" fallback={<Text text="Loading B" />}>
        <B />
      </Suspense>
      <Suspense key="1.1.c" fallback={<Text text="Loading C" />}>
        <C />
      </Suspense>
    </SuspenseList>
  </SuspenseList>
```

4. Resolve `A`

Given what I understood from `React.SuspenseList` I expect to see the following:
- After step 1: `Loading C`
- After step 2: `C`
- After step 3: `Loading A / Loading B / C` - because the key for C has not changed
- After step 4: `A / Loading B / C` - *but I have `Loading A / Loading B / C` instead*

**Comment by dubzzz** [comment]
@gaearon Is the scenario right above working as expected? Or should it be considered as an issue?

**Comment by kmurgic** [comment]
I'm having what I believe is a similar/closely related issue.

When using useTransition in combination with a SuspenseList with revealOrder="forwards" if some data is loaded during the transition it will render out of order.

Here is a codepen to demonstrate the issue:
https://codesandbox.io/s/awesome-brook-ykt49

If you click the next button sometimes the ProfileTimeline will render before the Profile Details (if the data for ProfileTimeline came back during the timeout for useTransition). @gaearon is this behavior intentional/expected?

It would be more intuitive if SuspenseList prevented its Suspense children from rendering out of order.  


```
const initialResource = fetchProfileData(0);

function getNextId(id) {
  return id === 3 ? 0 : id + 1;
}

function App() {
  const [resource, setResource] = useState(initialResource);
  const [startTransition, isPending] = useTransition({
    timeoutMs: 1000
  });

  return (
    <>
      <button
        disabled={isPending}
        onClick={() => {
          startTransition(() => {
            const nextUserId = getNextId(resource.userId);
            setResource(fetchProfileData(nextUserId));
          });
        }}
      >
        Next
      </button>
      <Suspense fallback={<h1>Loading...</h1>}>
        <ProfilePage resource={resource} />
      </Suspense>
    </>
  );
}

function ProfilePage({ resource }) {
  return (
    <SuspenseList tail="collapsed" revealOrder="forwards">
      <Suspense fallback={<h2>Loading profile...</h2>}>
        <ProfileDetails resource={resource} />
      </Suspense>
      <Suspense fallback={<h2>Loading posts...</h2>}>
        <ProfileTimeline resource={resource} />
      </Suspense>
    </SuspenseList>
  );
}
```

**Comment by dubzzz** [comment]
Closing the issue as the originally reported issue is an expected behaviour due to a lack of key in my code snippet.

I might re-open another issue concerning the second problem as it does not seem to be expected. But I first need to check if it is still the case with the latest version.
